File: simple-stack-example/src/main/java/com/zhuinden/simplestackdemoexample/demo/FirstView.java
package com.zhuinden.simplestackdemoexample.demo;

import android.annotation.TargetApi;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.widget.RelativeLayout;

import com.zhuinden.simplestack.Backstack;
import com.zhuinden.simplestack.Bundleable;
import com.zhuinden.simplestackdemoexample.R;

import butterknife.ButterKnife;
import butterknife.OnClick;
/**
 * Created by Owner on 2017. 01. 12..
 */

public class FirstView
        extends RelativeLayout
        implements Bundleable {
    private static final String TAG = "FirstView";

    public FirstView(Context context) {
        super(context);
        init(context);
    }

    public FirstView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public FirstView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    @TargetApi(21)
    public FirstView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        init(context);
    }

    Backstack backstack;

    FirstKey firstKey;

    private void init(Context context) {
        if(!isInEditMode()) {
            backstack = BackstackService.get(context);
            firstKey = Backstack.getKey(context);
        }
    }

    @OnClick(R.id.first_button)
    public void clickButton(View view) {
        backstack.goTo(new SecondKey());
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        ButterKnife.bind(this);
    }

    @Override
    public Bundle toBundle() {
        Bundle bundle = new Bundle();
        bundle.putString("HELLO", "WORLD");
        return bundle;
    }

    @Override
    public void fromBundle(@Nullable Bundle bundle) {
        if(bundle != null) {
            Log.i(TAG, bundle.getString("HELLO"));
        }
    }
}

File: simple-stack-example-basic-kotlin/src/main/kotlin/com/zhuinden/simplestackdemoexample/common/FirstView.kt
package com.zhuinden.simplestackdemoexample.common

import android.annotation.TargetApi
import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.view.View
import android.widget.RelativeLayout
import butterknife.ButterKnife
import butterknife.OnClick
import com.zhuinden.simplestack.Backstack
import com.zhuinden.simplestack.Bundleable
import com.zhuinden.simplestackdemoexample.R
import com.zhuinden.statebundle.StateBundle

/**
 * Created by Owner on 2017. 01. 12..
 */
class FirstView : RelativeLayout, Bundleable {
    constructor(context: Context) : super(context) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {
        init(context)
    }

    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int) : super(context, attrs, defStyleAttr) {
        init(context)
    }

    @TargetApi(21)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int, defStyleRes: Int) : super(context, attrs, defStyleAttr, defStyleRes) {
        init(context)
    }

    lateinit internal var backstack: Backstack

    lateinit internal var firstKey: FirstKey

    private fun init(context: Context) {
        if (!isInEditMode) {
            backstack = BackstackService.get(context)
            firstKey = Backstack.getKey<FirstKey>(context)
        }
    }

    @OnClick(R.id.first_button)
    fun clickButton(view: View) {
        backstack.goTo(SecondKey.create())
    }

    override fun onFinishInflate() {
        super.onFinishInflate()
        ButterKnife.bind(this)
    }

    override fun toBundle(): StateBundle {
        return StateBundle().apply {
            putString("HELLO", "WORLD")
        }
    }

    override fun fromBundle(bundle: StateBundle?) {
        bundle?.let {
            Log.i(TAG, it.getString("HELLO"))
        }
    }

    companion object {
        private val TAG = "FirstView"
    }
}
